### DISCUSSION TITLE
Integration of Publii with Cloudflare R2

### DISCUSSION THREAD

**Discussion by contano** [GetPublii/Publii#1782] [Ideas]
Hello,

Are you planning to build an integration of Publii with Cloudflare R2? Cloudflare R2 is a great alternative to Google Cloud and Amazon S3. The costs of hosting a website on Cloudflare R2 are significantly lower than on Amazon S3.

Regards

Jacek Cholewa

**Comment by dziudek** [comment]
Hi,

To be honest - I do not see any sense to use R2 if you have Cloudflare Pages which are totally free and dedicated for static pages hosting :)

**Reply by contano** [reply]
I created my photo galleries using Publii. I used the Ella template, and it looked nice. I configured it through GitHub + Cloudflare, and it worked well until the size of the galleries increased significantly. Problems started with uploading data to GitHub. I optimize the photos before uploading. I don't upload them carelessly. These photo galleries contain a lot of images, and honestly, it is not suitable for hosting on GitHub. I gave up using Publii for this purpose and moved the galleries to Adobe Portfolio. There, I don't have any problems, and everything looks nice and works well, but I don't have the same configuration and customization options as in Publii. I kept my blog on Publii, which doesn't require as many resources. I tested Publii with Amazon S3, and it works, but the prices in Cloudflare R2 are much better. Considering the needs of users who upload more resources, Cloudflare R2 is a good option, even a very good option in terms of price :)

Here is the link where you can find a price comparison between Amazon S3 and Cloudflare R2. A pricing calculator is also available.

https://www.cloudflare.com/pl-pl/lp/pg-cloudflare-r2-vs-aws-s3/?gad_source=1#htmlIdPlansMatrix

**Reply by contano** [reply]
![Zrzut ekranu 2024-11-15 220526](https://github.com/user-attachments/assets/4cd0416c-4774-4c10-a92d-bbd0ceb8384d)
Thank you for the suggestion. Yes, it works. I am now hosting my Publii sites with Cloudflare R2. It was necessary to set up a transformation rule in Cloudflare. I am attaching a screenshot of the rule because you might want to add it to the configuration methods in your tutorials someday. The R2 setup itself is incredibly simple. In my opinion, it's a fantastic option for Publii users.
Regards
Jacek

**Reply by akarshhemrajanime** [reply]
Can Publii directly connect to Cloudflare pages?

**Reply by contano** [reply]
The most convenient way to connect Publii to Cloudflare Pages is through GitHub. I used to do this in the past, but now I use Cloudflare R2. If you want to connect Publii directly to Cloudflare Pages, you can export the Publii project to your disk and then use the option to directly upload those resources to Cloudflare Pages. You can probably also use Wrangler, but I haven't done this myself, so I can't say exactly how to use Wrangler for this purpose.
![Zrzut ekranu 2024-11-16 175024](https://github.com/user-attachments/assets/310db14c-0cd3-40f1-88db-6adade43d4ff)
Coopilot or Gemini will guide you :) I am not a programmer, but with the help of AI I have done a lot myself :)

**Reply by akarshhemrajanime** [reply]
Thank you very much

**Comment by dziudek** [comment]
@contano - BTW did you tried to use S3 provider in Publii with enabled option "Use a custom S3 provider"? As I remember R2 supports the S3 API: https://developers.cloudflare.com/r2/api/s3/api/

**Reply by contano** [reply]
Not yet. Thanks for the info. I'll give it a try.

**Comment by shawngiese** [comment]
Wow, that was so easy! Now I can publish right to R2! By default the index.html is not automatically loaded. That is why the transform rule is essential to making it all work. Not sure if this is better than CloudFlare pages but it is smooth! Love Publii!

**Comment by banisterious** [comment]
This is great stuff, thank you for sharing, contano.

It seems that after enabling the transform rule using what you illustrated above in the screenshot, Cloudflare is throwing the "page isn't redirecting properly" error. I've removed cache/cookies and experimented with SSL (switching back and forth between Full and Full strict). Any pointers or ideas for things to try?

Here's to hoping that this gets added to the tutorials!

**Reply by contano** [reply]
I don't know how to help you unfortunately. This configuration works fine for me. Perhaps someone more experienced in technical matters can help you.

**Reply by shawngiese** [reply]
Here is my URL where I have been testing https://mydatascience.net. My domain is registered with them too so maybe they are doing something there. Using their free license package so you should not need their commercial license to do this.

My only difference from the screenshot is that I chose "Custom filter expression" on that page (scroll to the top of the page to see it.)
In my R2 bucket, make sure to activate the custom domain.

**Reply by banisterious** [reply]
Thanks to both of you. I think Custom filter expression is actually a requirement for the rule (in any case mine is the same). I'll just table this for now and circle back later when I have more time. I'm using Cloudflare's image hosting, which should be a decent compromise for a while until the site gets legs.

Speaking of Cloudflare image hosting, it'd be wonderful to see some sort of Publii integration with that (i.e., upload images to Cloudflare on the fly); perhaps I should put in an FR? Or maybe that's too big of an ask. At the moment I'm manually replacing image links with links to Cloudflare versions, which is totally fine for now, but it probably won't be sustainable once I begin to roll out some additional image gallery pages that I have planned.

**Reply by shawngiese** [reply]
maybe try first with just a simple HTML file to be sure it is not an issue on the Cloudflare side.

**Comment by ckyakr** [comment]
I'm using a new [CDN plugin](https://github.com/GetPublii/Publii/discussions/2107) which repoints the image URLs upon publish.  I have a gitignore file to ignore the output/media folder so it doesn't commit hundreds (or thousands) of images up into GitHub.

The only manual step is to copy the media folder up to Cloudflare R2.  I'm using CyberDuck currently because it's like an FTP app for R2, I can browse, move, rename, delete files/folders in R2 https://www.answeroverflow.com/m/1155472378139267144

My next step would is to write a script that would use Rclone to do the uploads to R2.  The script could either be manually run, or it would be nice to have it  monitor the output/media folder for changes, but wait a set amount of time before starting since Publii erases that folder on every publish.  This way only new images are uploaded and orphans would be deleted.

Publii needs the ability to [run commands after a publish](https://github.com/GetPublii/Publii/discussions/2124) occurs, then I wouldn't have to monitor the output/media folder for changes, I could have it just fire off the Rclone script.
